optimize/cmaes.go

```diff
@@ -0,0 +1,381 @@
+	} else if cma.pop < 0 {
+		panic("cma-es-chol: negative population size")
+	}
+	mu := cma.pop / 2
+	cma.weights = resize(cma.weights, mu)
+	for i := range cma.weights {
+		v := math.Log(float64(mu)+0.5) - math.Log(float64(i)+1)
+		cma.weights[i] = v
+	}
+	floats.Scale(1/floats.Sum(cma.weights), cma.weights)
+	cma.muEff = 0
+	for _, v := range cma.weights {
+		cma.muEff += v * v
+	}
+	cma.muEff = 1 / cma.muEff
+
+	n := float64(dim)
+	cma.cc = (4 + cma.muEff/n) / (n + 4 + 2*cma.muEff/n)
+	cma.cs = (cma.muEff + 2) / (n + cma.muEff + 5)
+	cma.c1 = 2 / ((n+1.3)*(n+1.3) + cma.muEff)
+	cma.cmu = math.Min(1-cma.c1, 2*(cma.muEff-2+1/cma.muEff)/((n+2)*(n+2)+cma.muEff))
+	cma.ds = 1 + 2*math.Max(0, math.Sqrt((cma.muEff-1)/(n+1))-1) + cma.cs
+	// E[chi] is taken from https://en.wikipedia.org/wiki/CMA-ES (there
+	// listed as E[||N(0,1)||]).
+	cma.eChi = math.Sqrt(n) * (1 - 1.0/(4*n) + 1/(21*n*n))
+
+	// Allocate memory for function data.
+	cma.xs = mat.NewDense(cma.pop, dim, nil)
+	cma.fs = resize(cma.fs, cma.pop)
+
+	// Allocate and initialize adaptive parameters.
+	cma.sigma = cma.InitStepSize
+	if cma.sigma == 0 {
+		cma.sigma = 0.3
+	} else if cma.sigma < 0 {
+		panic("cma-es-chol: negative initial step size")
+	}
+	cma.pc = resize(cma.pc, dim)
+	for i := range cma.pc {
+		cma.pc[i] = 0
+	}
+	cma.ps = resize(cma.ps, dim)
+	for i := range cma.ps {
+		cma.ps[i] = 0
+	}
+	cma.mean = resize(cma.mean, dim)
+	if cma.InitMean != nil {
+		if len(cma.InitMean) != dim {
+			panic("cma-es-chol: InitMean length mismatch")
+		}
+		copy(cma.mean, cma.InitMean)
+	}
+	if cma.InitCholesky != nil {
+		cma.chol = mat.Cholesky{}
+		cma.chol.Clone(cma.InitCholesky)
+	} else {
+		// Set the initial Cholesky to I.
+		b := mat.NewDiagonal(dim, nil)
+		for i := 0; i < dim; i++ {
+			b.SetSymBand(i, i, 1)
+		}
+		var chol mat.Cholesky
+		ok := chol.Factorize(b)
+		if !ok {
+			panic("cma-es-chol: bad cholesky. shouldn't happen")
+		}
+		cma.chol = chol
+	}
+
+	cma.evals = make([]int, cma.pop)
+	for i := 0; i < cma.pop; i++ {
+		cma.evals[i] = i
+	}
+
+	cma.bestX = resize(cma.bestX, dim)
+	cma.bestF = math.Inf(1)
+
+	t := min(tasks, cma.pop)
+	cma.taskIdxs = make([]int, cma.pop)
+	for i := 0; i < t; i++ {
+		cma.taskIdxs[i] = -1
+	}
+	// Get a new waitgroup so that if the struct is re-used the.
+	cma.mux = sync.Mutex{}
+	cma.wg = sync.WaitGroup{}
+	return t
+}
+
+func (cma *CmaEsChol) IterateGlobal(task int, loc *Location) (Operation, error) {
+	// Check the status of the incoming task. If it is a number, it means
+	// that taks contains a valid location.
+	idx := cma.taskIdxs[task]
+	if idx != -1 {
+		cma.fs[idx] = loc.F
+		cma.wg.Done()
+	}
+
+	// Get the next task and send it to be run if there is a next task to be run.
+	// If all of the tasks have been run, perform an update step. Note that the
+	// use of this mutex means that only one task can proceed, all of the
+	// other tasks should get stuck and then get a new location.
+	cma.mux.Lock()
```

Comment:
I think I'd be inclined to use a `defer cma.mux.Unlock()` here.
(and remove the direct calls at lines 261 and 301)
what do you think?

Reply:
I thought about it, but there's a few lines in the `if` statement that don't have to be protected by a mutex, and relative to the part that needs to be protected, the part that doesn't need to be protected is expensive. It doesn't really matter that much though, because this algorithm is either going to be dominated by the function evaluations, or it's going to be dominated by the parameter update, so if you'd like the extra safety I can change to defer.

Reply:
yeah, I think I'd prefer with `defer`. (I strive to follow the rules I teach my students: first the correct, robust answer. then, faster :P)

Reply:
Actually, I do think it's better as is. The piece in the mutex in the short path is a very small constant (slice access and slice). The piece outside the mutex is a `dim^2` and then a `dim`. I'd like to keep that outside.

Reply:
There is a safer alternative, which is to place a defer after the condition here. It least its visible on the same page then when it's being read.